cmd/snapd-generator/main.c

```diff
@@ -127,6 +127,52 @@ static bool executable_exists(const char *name) {
 	return false;
 }
 
+static bool is_snap_try_snap_unit(const char *units_dir, const char *mount_unit_name) {
+	char fname[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(fname, sizeof fname, "%s/%s", units_dir, mount_unit_name);
+	FILE *f SC_CLEANUP(sc_cleanup_file) = NULL;
+	f = fopen(fname, "r");
+	if (!f) {
+		// not really expected
+		fprintf(stderr, "cannot open mount unit %s: %m\n", fname);
+		return false;
+	}
+
+	fseek(f, 0, SEEK_END);
+	long size = ftell(f);
+	fseek(f, 0, SEEK_SET);
+
+	// read entire mount unit file into buffer
+	char *buffer = NULL;
+	sc_cleanup_string(&buffer);
+	buffer = malloc(size + 1);
+	size_t n = fread(buffer, size, 1, f);
```

Comment:
why not a snippet like this:
```
char *line SC_CLEANUP(sc_cleanup_string) = NULL;
size_t line_buf_size = 0;
while(1) {
    errno = 0;
    ssize_t num_read = getline(&line, &line_buf_size, f);
    if (num_read < 0 && errno != 0) {
      // bail
    }
    if (line[num_read-1] == '\n') {
        line[num_read-1] = '\0';
    }
    ....
    const char *what_prefix = "What=";
    char *what = strstr(line, what_prefix);
    if (NULL == what) {
        continue;
    }
    char *start = what + strlen(what_prefix);
    if (strlen(start) == 0) {
        // bogus path?
        continue;
    }
    // always return beyond this point
    struct stat st;
    if (stat(start, &st) == 0 && (st.st_mode & S_IFMT) == S_IFDIR) {
        return true;
    }
    return false;
}
return false
```

and then you can also drop fseek()/ftell()

Reply:
This whole function can be replaced with `sc_infofile_get_key`, because ini-like systemd units still follow key=value pairs on individual lines.

Reply:
yup, @zyga's suggestion is definitely better!

Reply:
> This whole function can be replaced with `sc_infofile_get_key`, because ini-like systemd units still follow key=value pairs on individual lines.

Ouch, unfortunately this helper chokes on ini groups ([Unit] etc), it expects key=value only. I'm not sure this bugfix warrants extending sc_infofile_get_key - properly supporting groups there (instead of just skipping them) is probably a bigger change on its own.

Reply:
Implemented new sc_infofile_get_ini_key function for getting a key from ini sections and simplified the main code, please take a look.